File: src/bitsnap/http/headers/accept.kt
/**
 *  Copyright 2016 Yuriy Yarosh
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 **/

package bitsnap.http.headers

import bitsnap.exceptions.*
import bitsnap.http.*
import java.util.*
import java.util.Date

class Accept internal constructor(val types: List<Pair<MimeType, Int>>) : Header() {

    constructor(type: MimeType) : this(listOf(Pair(type, 10)))

    override val name = Accept.Companion.name

    override val value: String by lazy {
        joinToStringWithQuality(types) {
            it.typeToString()
        }
    }

    override fun equals(other: Any?) = if (other is Accept) {
        types == other.types
    } else false

    data class Builder internal constructor(private val types: MutableList<Pair<MimeType, Int>> = LinkedList()) {

        private fun type(type: Pair<MimeType, Int>) = checkQuality(Accept.name, type.second) {
            if (types.firstOrNull() { it.first == type.first } != null) {
                throw AlreadyAssignedException("MimeType ${type.first.typeToString()}")
            } else types.add(type)
        }

        fun type(type: MimeType, quality: Int) = type(Pair(type, quality))

        fun type(type: MimeType) = type(Pair(type, 10))

        fun types(vararg types: MimeType) = types.forEach {
            type(Pair(it, 10))
        }

        companion object {

            @Throws(HeaderDuplicateException::class)
            operator fun invoke(init: Builder.() -> Unit): Accept {
                val builder = Builder()
                builder.init()
                builder.types.checkHeaderPairDuplicates(Accept.name)
                return Accept(builder.types
                    .sortedWith(acceptQualityComparator)
                )
            }
        }
    }

    companion object : HeaderCompanion {

        init {
            Header.registerCompanion(Accept.Companion)
        }

        override val name = "Accept"

        override operator fun invoke(value: String) = Accept(value.splitValueQuality().map {
            val (stringValue, quality) = it
            Pair(MimeType(stringValue), quality)
        }.toList())

        @Throws(HeaderDuplicateException::class)
        operator fun invoke(init: Builder.() -> Unit) = Builder(init)
    }
}

class AcceptCharset internal constructor(val charsets: List<Pair<Charset, Int>>) : Header() {

    constructor(charset: Charset) : this(listOf(Pair(charset, 10)))

    override val name = AcceptCharset.Companion.name

    override val value: String by lazy {
        joinToStringWithQuality(charsets)
    }

    override fun equals(other: Any?) = if (other is AcceptCharset) {
        charsets == other.charsets
    } else false

    data class Builder internal constructor(private val charsets: MutableList<Pair<Charset, Int>> = LinkedList()) {

        private fun charset(charset: Pair<Charset, Int>) = checkQuality(AcceptCharset.name, charset.second) {
            if (charsets.firstOrNull() { it.first == charset.first } != null) {
                throw AlreadyAssignedException("Charset ${charset.first.toString()}")
            } else charsets.add(charset)
        }

        fun charset(charset: Charset, quality: Int) = charset(Pair(charset, quality))

        fun charset(charset: Charset) = charset(charset, 10)

        fun charsets(vararg charsets: Charset) = charsets.forEach {
            charset(Pair(it, 10))
        }

        companion object {

            @Throws(HeaderDuplicateException::class)
            operator fun invoke(init: Builder.() -> Unit): AcceptCharset {
                val builder = Builder()
                builder.init()
                builder.charsets.checkHeaderPairDuplicates(AcceptCharset.name)
                return AcceptCharset(builder.charsets
                    .sortedWith(acceptQualityComparator)
                )
            }
        }
    }

    companion object : HeaderCompanion {

        init {
            Header.registerCompanion(AcceptCharset.Companion)
        }

        override val name = "Accept-Charset"

        @Throws(UnknownCharsetException::class)

        override operator fun invoke(value: String) = AcceptCharset(value.splitValueQuality().map {
            Pair(Charset(it.first), it.second)
        }.toList())

        @Throws(HeaderDuplicateException::class)
        operator fun invoke(init: Builder.() -> Unit) = Builder(init)
    }
}

class AcceptEncoding internal constructor(val types: List<Pair<Encoding, Int>>) : Header() {

    constructor(type: Encoding) : this(listOf(Pair(type, 10)))

    override val name = AcceptEncoding.Companion.name

    override val value: String by lazy {
        joinToStringWithQuality(types)
    }

    override fun equals(other: Any?) = if (other is AcceptEncoding) {
        types == other.types
    } else false

    data class Builder internal constructor(private val encodings: MutableList<Pair<Encoding, Int>> = LinkedList()) {

        private fun encoding(encoding: Pair<Encoding, Int>) = checkQuality(AcceptEncoding.name, encoding.second) {
            if (encodings.firstOrNull() { it.first == encoding.first } != null) {
                throw AlreadyAssignedException("Encoding ${encoding.first.toString()}")
            } else encodings.add(encoding)
        }

        fun encoding(encoding: Encoding, quality: Int) = encoding(Pair(encoding, quality))

        fun encoding(encoding: Encoding) = encoding(encoding, 10)

        fun encodings(vararg encodings: Encoding) = encodings.forEach {
            encoding(Pair(it, 10))
        }

        companion object {
            @Throws(HeaderDuplicateException::class)
            operator fun invoke(init: Builder.() -> Unit): AcceptEncoding {
                val builder = Builder()
                builder.init()
                builder.encodings.checkHeaderPairDuplicates(AcceptEncoding.name)
                return AcceptEncoding(builder.encodings
                    .sortedWith(acceptQualityComparator)
                )
            }
        }
    }

    companion object : HeaderCompanion {

        init {
            Header.registerCompanion(AcceptEncoding.Companion)
        }

        override val name = "Accept-Encoding"

        @Throws(UnknownEncodingException::class)
        override operator fun invoke(value: String): AcceptEncoding {
            return AcceptEncoding(value.splitValueQuality().map {
                val (encodingTypeValue, quality) = it
                Pair(Encoding(encodingTypeValue), quality)
            }.toList())
        }

        @Throws(HeaderDuplicateException::class)
        operator fun invoke(init: Builder.() -> Unit) = Builder(init)
    }
}

class AcceptLanguage internal constructor(val locales: List<Pair<Locale, Int>>) : Header() {

    constructor(locale: Locale) : this(listOf(Pair(locale, 10)))

    override val name = AcceptLanguage.Companion.name

    override val value: String by lazy {
        joinToStringWithQuality(locales)
    }

    override fun equals(other: Any?) = if (other is AcceptLanguage) {
        locales == other.locales
    } else false

    data class Builder internal constructor(private val locales: MutableList<Pair<Locale, Int>> = LinkedList()) {

        private fun locale(locale: Pair<Locale, Int>) = checkQuality(AcceptLanguage.name, locale.second) {
            if (locales.firstOrNull() { it.first == locale.first } != null) {
                throw AlreadyAssignedException("Encoding ${locale.first.toString()}")
            } else locales.add(locale)
        }

        fun locale(locale: Locale, quality: Int) = locale(Pair(locale, quality))

        fun locale(locale: Locale) = locale(locale, 10)

        fun locales(vararg locales: Locale) = locales.forEach {
            locale(Pair(it, 10))
        }

        companion object {

            @Throws(HeaderDuplicateException::class)
            operator fun invoke(init: Builder.() -> Unit): AcceptLanguage {
                val builder = Builder()
                builder.init()
                builder.locales.checkHeaderPairDuplicates(AcceptLanguage.name)
                return AcceptLanguage(builder.locales
                    .sortedWith(acceptQualityComparator)
                )
            }
        }
    }

    companion object : HeaderCompanion {

        init {
            Header.registerCompanion(AcceptLanguage.Companion)
        }

        override val name = "Accept-Language"

        override operator fun invoke(value: String): AcceptLanguage {
            return AcceptLanguage(value.splitValueQuality().map {
                val (stringValue, quality) = it
                val sepIndex = stringValue.indexOfAny(listOf("-", "_"))

                if (sepIndex > 0) {
                    val language = stringValue.substring(0, sepIndex)
                    val country = stringValue.substring(sepIndex + 1)
                    Pair(Locale(language, country), quality)
                } else {
                    Pair(Locale(stringValue), quality)
                }
            }.toList())
        }

        @Throws(HeaderDuplicateException::class)
        operator fun invoke(init: Builder.() -> Unit) = Builder(init)
    }
}

class AcceptDatetime(val date: Date) : Header() {

    override val name = AcceptDatetime.Companion.name

    override val value: String by lazy {
        formatDate(date)
    }

    override fun equals(other: Any?) = if (other is AcceptDatetime) {
        date.equals(other.date)
    } else false

    companion object : HeaderCompanion {

        init {
            Header.registerCompanion(AcceptDatetime.Companion)
        }

        override val name = "Accept-Datetime"

        @Throws(HeaderDateParseException::class)
        override operator fun invoke(value: String) = AcceptDatetime(parseDate(value))
    }
}

class AcceptPatch internal constructor(val types: List<MimeType>) : Header() {

    constructor(type: MimeType) : this(listOf(type))

    override val name = AcceptPatch.Companion.name

    override val value: String by lazy {
        types.joinToString(", ")
    }

    override fun equals(other: Any?) = if (other is AcceptPatch) {
        types == other.types
    } else false

    data class Builder internal constructor(private val types: MutableList<MimeType> = LinkedList()) {

        fun type(type: MimeType) = types.add(type)

        fun types(vararg types: MimeType) = this.types.addAll(types)

        companion object {
            operator fun invoke(init: Builder.() -> Unit): AcceptPatch {
                val builder = Builder()
                builder.init()
                builder.types.checkHeaderDuplicates(AcceptPatch.name)
                return AcceptPatch(builder.types)
            }
        }
    }

    companion object : HeaderCompanion {

        init {
            Header.registerCompanion(AcceptPatch.Companion)
        }

        override val name = "Accept-Patch"

        override operator fun invoke(value: String) = AcceptPatch(value.split(", ").map { MimeType(it) })

        operator fun invoke(init: Builder.() -> Unit) = Builder(init)
    }
}

class AcceptRanges(val unit: RangeUnit) : Header() {

    override val name = AcceptRanges.Companion.name

    override val value: String by lazy {
        unit.toString()
    }

    override fun equals(other: Any?) = if (other is AcceptRanges) {
        unit == other.unit
    } else false

    companion object : HeaderCompanion {

        init {
            Header.registerCompanion(AcceptRanges.Companion)
        }

        override val name = "Accept-Ranges"

        override operator fun invoke(value: String) = AcceptRanges(RangeUnit(value))
    }
}

File: src/http/headers/Accept.scala
/**
  *  Copyright 2016 Yuriy Yarosh
  *
  *  Licensed under the Apache License, Version 2.0 (the "License");
  *  you may not use this file except in compliance with the License.
  *  You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  *  Unless required by applicable law or agreed to in writing, software
  *  distributed under the License is distributed on an "AS IS" BASIS,
  *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  *  See the License for the specific language governing permissions and
  *  limitations under the License.
  **/
package io.bitsnap.http
package headers

import java.util.Locale

import Header.{Date => HeaderDate, _}
import io.bitsnap.http.{Range => HttpRange}
import Header.Implicits._

import scala.util.{Failure, Success, Try}

class Accept(val types: QualityParameters[MimeType]) extends Header {

  if (types.isEmpty) {
    throw Accept.Invalid
  }

  override lazy val value = types.toString

  override val name = Accept.name

  def this(`type`: MimeType) {
    this(Seq(`type` -> 10))
  }

  override def equals(that: Any) = that match {
    case that: Accept => this.types == that.types
    case _            => false
  }

  override def hashCode = (0 /: types.map { _.hashCode }) { headers.hashCodePrime + _ + _ }
}

object Accept {

  private[http] final val name = "accept"

  override def toString = name

  object Invalid extends Header.Invalid

  def apply(string: String): Try[Accept] = {
    if (string.trim.isEmpty) {
      Failure(Invalid)
    } else {
      val Parameters(parameters @ _ *) = string

      val qualityParameters = parameters.map {
        _.toQualityParameter(MimeType(_), Invalid)
      }.filter { e =>
        e.isSuccess && e.get._1.isSuccess
      }.map { e =>
        (e.get._1.get, e.get._2)
      }

      qualityParameters.isEmpty match {
        case true => Failure(Invalid)
        case _    => Success(new Accept(qualityParameters))
      }
    }
  }
}

final class AcceptCharset(val charsets: QualityParameters[Charset]) extends Header {

  if (charsets.isEmpty) {
    throw AcceptCharset.Invalid
  }

  override val name       = AcceptCharset.name
  override lazy val value = charsets.toString

  def this(charset: Charset) {
    this(Seq(charset -> 10))
  }

  override def equals(that: Any) = that match {
    case that: AcceptCharset => this.charsets == that.charsets
    case _                   => false
  }

  override def hashCode = (0 /: charsets.map { _.hashCode }) { headers.hashCodePrime + _ + _ }
}

object AcceptCharset {

  private[http] final val name = "accept-charset"

  override def toString = name

  object Invalid extends Header.Invalid

  def apply(string: String): Try[AcceptCharset] = {
    if (string.trim.isEmpty) {
      Failure(Invalid)
    } else {
      val Parameters(parameters @ _ *) = string

      val qualityParameters = parameters.map {
        _.toQualityParameter(Charset(_), Invalid)
      }.filter { e =>
        e.isSuccess && e.get._1.isSuccess
      }.map { e =>
        (e.get._1.get, e.get._2)
      }

      qualityParameters.isEmpty match {
        case true => Failure(Invalid)
        case _    => Success(new AcceptCharset(qualityParameters))
      }
    }
  }
}

final class AcceptDatetime(override val date: HeaderDate) extends DateHeader(date) {
  override val name = AcceptDatetime.name

  override def equals(that: Any) = that match {
    case that: AcceptDatetime => super.equals(that)
    case _                    => false
  }

  override def hashCode = headers.hashCodePrime + super.hashCode
}

object AcceptDatetime {

  private[http] final val name = "accept-datetime"

  override def toString = name

  object Invalid extends Header.Invalid

  def apply(string: String): Try[AcceptDatetime] = {
    if (string.trim.isEmpty) {
      Failure(Invalid)
    } else {
      HeaderDate(string).map { new AcceptDatetime(_) }
    }
  }
}

final class AcceptEncoding(val encodings: QualityParameters[Encoding]) extends Header {

  // Trailers restricted only for use in TE header
  if (encodings.isEmpty || encodings.contains(Encoding.Trailers)) {
    throw AcceptEncoding.Invalid
  }

  override val name       = AcceptEncoding.name
  override lazy val value = encodings.toString

  def this(encoding: Encoding) {
    this(Seq(encoding -> 10))
  }

  override def equals(that: Any) = that match {
    case that: AcceptEncoding => this.encodings == that.encodings
    case _                    => false
  }

  override def hashCode = (0 /: encodings.map { _.hashCode }) { headers.hashCodePrime + _ + _ }
}

object AcceptEncoding {

  private[http] final val name = "accept-encoding"

  override def toString = name

  object Invalid extends Header.Invalid

  def apply(string: String): Try[AcceptEncoding] = {
    if (string.trim.isEmpty) {
      Failure(Invalid)
    } else {

      val Parameters(parameters @ _ *) = string

      val qualityParameters = parameters.map {
        _.toQualityParameter(Encoding(_), Invalid)
      }.filter { e =>
        e.isSuccess && e.get._1.isSuccess
      }.map { e =>
        (e.get._1.get, e.get._2)
      }

      qualityParameters.isEmpty match {
        case true => Failure(Invalid)
        case _    => Success(new AcceptEncoding(qualityParameters))
      }
    }
  }
}

final class AcceptLanguage(val languages: QualityParameters[Locale]) extends Header {

  if (languages.isEmpty) {
    throw AcceptLanguage.Invalid
  }

  override val name       = AcceptLanguage.name
  override lazy val value = languages.toString

  def this(language: Locale) {
    this(Seq(language -> 10))
  }

  override def equals(that: Any) = that match {
    case that: AcceptLanguage => this.languages == that.languages
    case _                    => false
  }

  override def hashCode =
    (0 /: languages.map { _.hashCode }) { headers.hashCodePrime + _ + _ }
}

object AcceptLanguage {

  private[http] final val name = "accept-language"

  override def toString = name

  object Invalid extends Header.Invalid

  def apply(string: String): Try[AcceptLanguage] = {
    if (string.trim.isEmpty) {
      Failure(Invalid)
    } else {
      val Parameters(parameters @ _ *) = string

      val qualityParameters = parameters.map { p =>
        p.toQualityParameter({ e =>
          val Locales(locale) = e; locale
        }, Invalid)
      }.filter { _.isSuccess }.map { _.get }

      qualityParameters.isEmpty match {
        case true => Failure(Invalid)
        case _    => Success(new AcceptLanguage(qualityParameters))
      }
    }
  }
}

final class AcceptPatch(types: QualityParameters[MimeType]) extends Accept(types) {

  if (types.isEmpty) {
    throw AcceptPatch.Invalid
  }

  override val name = AcceptPatch.name

  def this(`type`: MimeType) {
    this(Seq(`type` -> 10))
  }

  override def equals(that: Any) = that match {
    case that: AcceptPatch => super.equals(that)
    case _                 => false
  }

  override def hashCode = headers.hashCodePrime + super.hashCode
}

object AcceptPatch {

  private[http] final val name = "accept-patch"

  override def toString = name

  object Invalid extends Header.Invalid

  def apply(string: String): Try[AcceptPatch] = {
    if (string.isEmpty) {
      Failure(Invalid)
    } else {
      val Parameters(parameters @ _ *) = string

      val qualityParameters = parameters.map {
        _.toQualityParameter(MimeType(_), Invalid)
      }.filter { e =>
        e.isSuccess && e.get._1.isSuccess
      }.map { e =>
        (e.get._1.get, e.get._2)
      }

      qualityParameters.isEmpty match {
        case true => Failure(Invalid)
        case _    => Success(new AcceptPatch(qualityParameters))
      }
    }
  }
}

final class AcceptRanges(val unit: HttpRange.Unit) extends Header {

  override lazy val value = unit.toString
  override val name       = AcceptRanges.name

  override def equals(that: Any) = that match {
    case that: AcceptRanges => this.unit == that.unit
    case _                  => false
  }

  override def hashCode = headers.hashCodePrime + unit.hashCode
}

object AcceptRanges {

  private[http] final val name = "accept-ranges"

  override def toString = name

  object Invalid extends Header.Invalid

  def apply(string: String): Try[AcceptRanges] = {
    if (string.isEmpty || string.trim.isEmpty) {
      Failure(Invalid)
    } else {
      HttpRange.Unit(string).map { new AcceptRanges(_) }
    }
  }
}

final class TransferEncodings(val encodings: Seq[Encoding]) extends Header {
  override val name: String = TransferEncodings.name

  override lazy val value: String = encodings.mkString(", ")

  override def equals(that: Any) = that match {
    case that: TransferEncodings => this.encodings == that.encodings
    case _                       => false
  }

  override def hashCode = (0 /: encodings.map { _.hashCode }) { headers.hashCodePrime + _ + _ }
}

object TransferEncodings {

  private[http] final val name = "te"

  override def toString = name

  object Invalid extends Header.Invalid

  def apply(string: String): Try[TransferEncodings] =
    if (string.trim.isEmpty) {
      Failure(Invalid)
    } else {
      val Encoding(encodings @ _ *) = string

      encodings.isEmpty match {
        case true => Failure(Invalid)
        case _    => Success(new TransferEncodings(encodings))
      }
    }
}
